airflow_clickhouse_plugin/sensors/clickhouse_sql_sensor.py

```diff
@@ -0,0 +1,73 @@
+from typing import Dict, Callable, Any
+
+from airflow.exceptions import AirflowException
+from airflow.sensors.sql_sensor import SqlSensor
+from airflow_clickhouse_plugin.hooks.clickhouse_hook import ClickHouseHook
+
+
+class ClickHouseSqlSensor(SqlSensor):
+    def __init__(
+        self,
+        sql: str = None,
+        clickhouse_conn_id: str = 'clickhouse_default',
+        parameters: Dict[str, Any] = None,
+        database: str = None,
+        success: Callable[[Any], bool] = None,
+        failure: Callable[[Any], bool] = None,
+        fail_on_empty: bool = False,
+        *args,
+        **kwargs,
+    ):
+        super().__init__(
+            conn_id=clickhouse_conn_id,
+            sql=sql,
+            parameters=parameters,
+            success=success,
+            failure=failure,
+            fail_on_empty=fail_on_empty,
+            *args,
+            **kwargs,
+        )
+        self._database = database
+
+    def poke(self, context):
```

Comment:
I have finally got an idea how to do it in a least destructive way! Please do as follows:
```python
def __poke_backport(self, context):  # private method
    # your current code of the method including link to source

def poke(self, context):
    try:
        return super().poke(context)
    except AttributeError:  # fallback for earlier versions
        return self.__poke_backport(context):
```

This is like handling `ImportError` for different versions of Python in case if the package changed its name, for instance. Here is an example for better understanding:
https://github.com/aio-libs/aiohttp/blob/159b039b742aa86404ba9a66873a031d16e595e6/aiohttp/locks.py#L5-L8

If the further version of Airflow provides a better implementation, the plugin will use it instead of out backport.

Reply:
I also highly appreciate running tests for both versions which defines `poke` method (Airflow 1.10.8 if I am not mistaken)  and does not. But event better way is to create the mock tests with patched method. Do you need any help with that? Two tests needed:
1. Explicitly delete the method from the superclass definition (imitation of `AttributeError`) and test that our `__poke_backport` (which is also a patched mock) is called.
2. Explicitly add `poke` method to the superclass (as a mock) and test that it is called and `__poke_backport` is not.

Here are some docs: https://docs.python.org/3/library/unittest.mock.html#unittest.mock.patch

Reach out to me if you need any help with tests, I will implement them myself. Hope you are ok with such a long work on this method: we have to follow best practices and fight the reasons not the symptoms :) I believe this last approach is the most correct since we finally solve the exact issue.

Reply:
BTW please also add annotations for the `poke` and backport: `-> bool` and, if I am not mistaken, `context: dict` (better check yourself). According to your tests, it has to be `Optional[dict]`.

Reply:
@bryzgaloff I really appreciate your feedback, thank you. `poke` method is the method of base class BaseSensorOperator and should be overridden in descendant classes so checking `super().poke(context)` won't work. From the other side, starting from Airflow [1.10.7](https://github.com/apache/airflow/blob/1.10.7/airflow/sensors/sql_sensor.py) `SqlSensor` has `_get_hook` which makes the implementation of custom sensor mutch easier, like:
```python
    def _get_hook(self):
        return ClickHouseHook(clickhouse_conn_id=self.conn_id, database=self._database)
```
with no need to override the `poke` method.

I see that it could be implemented as follow:
```python
    def poke(self, context):
        if hasattr(super(), '_get_hook'):
            return super().poke(context)
        else:
            return self.__poke_backport(context)
```
Is it a good idea?

Reply:
Oh, yep, thank you for reminding the context, I have totally forgotten about the initial issue.

So to handle the case your idea looks good. But I'd rather recommend to solve the exact issue and in fact compare the version of installed to 1.10.7 when `_get_hook` was added. Since the presence of the method does not guarantee it is called.

I have looked for several ways on how to compare versions using stdlib of Python. The best way seems to be using `setuptools.extern.packaging.version.parse`. So I would recommend you using `version.parse(airflow.__version__) >= version.Version('1.10.7')`. If the condition is true then call `super().poke()`. And of course please call the exact implemented `_get_hook` method inside your `__poke_backport`. `packaging` [is said to be vendored](https://stackoverflow.com/a/11887885/11222662) into `setuptools` so should be hopefully present on every Python installation.

But as a fallback I would recommend to also check for `ImportError` (if `setuptools` version is too old or `packaging` is not vendored). And in case of exception use `distutils.version.LooseVersion(airflow.__version__) >= LooseVersion('1.10.7')` ([docs](https://hg.python.org/cpython/file/tip/Lib/distutils/version.py#l267)). The reason why I do not recommend to use this solution as the primary one is that it treats release candidates to be ≥ than the release version:
```python
>>> LooseVersion('1.10.7rc1') >= LooseVersion('1.10.7')
True
```
So it can produce incorrect results for such versions. And if it fails too than we just unconditionally fallback to our backport method. `version.parse` has no such problem:
```python
>>> packaging.version.parse('0.1.1rc1') < packaging.version.parse('0.1.1')
True
```

I hope this will not be much harder to test since `airflow.__version__` is hopefully patchable either. Or the versions parser can be patched. I would also recommend encapsulating this versions comparing logic into a separate function inside sensor's module for better readability. And some tests for the function are also highly appreciated. Hopefully you wil be able to imitate `ImportError` to test fallback path too but if not then reach out to me, we will look for a solution. Does it sound reasonable?

Please also notice that `hasattr(super(), …)` will raise an exception, if I am not mistaken :)

Reply:
Just asking to be clear, `_get_hook` method was [intentionally added to customize subclasses](https://github.com/apache/airflow/commit/3c9e2ef140ba0ea7cbdc28bae64db81bf0286535) and is it a real situation when it exists and not called? Also, `hasattr(super(), …)` approach works well, at least on 1.10.6 and 1.10.10 :)
So checking versions by several methods with exceptions seems overcomplicated. What do you think?

Reply:
The reason why I recommended to use versions comparison is that it is quite straightforward and expresses exactly what we are trying to solve.

But since it is really too complicated to implement, I agree to follow the simplest way with `hasattr` at this iteration. Please just leave a descriptive comment there for the reader to fully understand the origin of the approach.